### PR TITLE
#1947: Create new method in DiscoveryServiceRegistry to expose discovery service timeout

### PR DESCRIPTION
**Pull request by ivivanov-bg**

### CODE REVIEW

**Comment by sjka** [comment]
lgtm

**Comment by kaikreuzer** [comment]
Thanks @ivivanov-bg, looks good.
What I would like to ask you is to also make this feature available through the REST API by adapting the DiscoveryResource class. We could then ask @aounhaider1 to add support in the Paper UI for it - this would make sure that the feature is actively used in ESH and that possible problems are seen early on.

**Comment by ivivanov-bg** [comment]
OK - I have an idea:
  Create one more method in `DiscoveryServiceRegistry`: `getSupportedBindingsWithScanTimeout` which will return a `Map` with key : value `bindingId` : `maxScanTimeout` 

And later - create new endpoint in `DiscoveryResource` to retrieve this data.

What do you think ?

**Comment by kaikreuzer** [comment]
I am not sure who would call the method getSupportedBindingsWithScanTimeout - what is its use case?
I think it is good enough to have '/rest/discovery/bindings/{bindingId}/timeout' returning the timeout on a HTTP GET. This could be called by the Paper UI to know how long the scan will take.

Maybe an even nicer idea: We could directly modify `/rest/discovery/bindings/{bindingId}/scan` to not simply return HTTP OK, but to return the timeout as a result - in this case, the Paper UI would not have to do a second HTTP request in order to get hold of that value. What do you think about that idea?

**Comment by ivivanov-bg** [comment]
I wanted to avoid the need of second HTTP call to get the timeout (the case with the new endpoint '/rest/discovery/bindings/{bindingId}/timeout')

That's why I came up with the idea to have an endpoint which will return both: bindingId and timeout with a single call.

However - your idea (having the `/rest/discovery/bindings/{bindingId}/scan` returning the timeout in the response) is maybe the best option: it should not break anything and anyone can use the payload if he wants

**Comment by kaikreuzer** [comment]
go for it then 😎

**Comment by ivivanov-bg** [comment]
Done !

File: bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/DiscoveryServiceRegistry.java [hunk 1]
```diff
@@ -21,6 +21,7 @@
  * </ul>
  *
  * @author Michael Grammling - Initial Contribution.
+ * @author Ivaylo Ivanov <ivivanov.bg@gmail.com> - Added getMaxScanTimeout
```
Comment (kaikreuzer) [inline bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/DiscoveryServiceRegistry.java]: please remove your e-mail from the author tags (to be in sync with the other entries)

Comment (ivivanov-bg) [inline bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/DiscoveryServiceRegistry.java]: done

File: bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/discovery/DiscoveryResource.java [hunk 2]
```diff
@@ -88,7 +89,23 @@ public void onFinished() {
...
+    }
+    
+    /**
+     * Dummy class needed to create a proper JSON formating of the response
+     * 
+     * @author ivaylo.ivanov
+     *
+     */
+    @SuppressWarnings("unused")
+    private static final class ScanResponseResultDto {
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/discovery/DiscoveryResource.java]: Do we really need an DTO class for this?
How about simply returning the integer as PLAIN? @aounhaider1 would that do or is it always preferred to receive JSON, even for single values?

File: bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/discovery/DiscoveryResource.java [hunk 3]
```diff
@@ -88,7 +89,23 @@ public void onFinished() {
                 logger.debug("Scan for binding '{}' successfully finished.", bindingId);
             }
         });
-        return Response.ok().build();
+        
+        return Response.ok(new ScanResponseResultDto(discoveryServiceRegistry.getMaxScanTimeout(bindingId))).build();
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/discovery/DiscoveryResource.java]: You have to adapt the Swagger documentation in line 78 as well.

File: bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/discovery/DiscoveryResource.java (hunk 2 above)
Comment (ivivanov-bg) [inline bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/discovery/DiscoveryResource.java]: Yes, we do. 

I tried simply returning the integer, but then - the response was empty.

Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/discovery/DiscoveryResource.java]: Then you did something wrong. It is of course possible to return PLAIN values as an HTTP response, we are e.g. doing it [here](https://github.com/eclipse/smarthome/blob/master/bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/item/ItemResource.java#L205).

Comment (ivivanov-bg) [inline bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/discovery/DiscoveryResource.java]: Ok - found the issue - it's fixed now

File: bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/discovery/DiscoveryResource.java (hunk 3 above)
Comment (ivivanov-bg) [inline bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/discovery/DiscoveryResource.java]: Done

File: bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/discovery/DiscoveryResource.java [hunk 4]
```diff
@@ -73,7 +74,8 @@ public Response getDiscoveryServices() {
 
     @POST
     @Path("/bindings/{bindingId}/scan")
-    @ApiOperation(value = "Starts asynchronous discovery process for a binding.")
+    @Produces(MediaType.TEXT_PLAIN)
+    @ApiOperation(value = "Starts asynchronous discovery process for a binding and returns the timeout of the discovery operation.", response = Integer.class)
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/discovery/DiscoveryResource.java]: "timeout in seconds" makes it probably clearer what exactly to expect :-)

Comment (ivivanov-bg) [inline bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/discovery/DiscoveryResource.java]: Done

**Comment by kaikreuzer** [comment]
Thanks :-)

**Comment by kaikreuzer** [comment]
I have created https://github.com/eclipse/smarthome/issues/1975 as a follow-up for the Paper UI for @aounhaider1 to work on.

**Comment by cdjackson** [comment]
This is a useful feature, but it would also be useful to have notifications when the discovery starts and stops. In Zwave, discovery will stop once the first device is discovered, so knowing the static time doesn't guarantee that the UI will correctly represent this information.

This is the initial use case for the binding notifications I proposed (https://github.com/openhab/openhab2-addons/pull/977), or in the notification system (https://github.com/eclipse/smarthome/pull/459).

Anyway - this is just for information - nothing wrong with this PR :)

**Comment by cdjackson** [comment]
I propose to change the response to use json -: https://github.com/eclipse/smarthome/pull/1978

**Pull request merged**